source/pervasive/seq_lib.rs

```diff
@@ -18,6 +18,54 @@ impl<A> Seq<A> {
         Seq::new(self.len(), |i: int| f(i, self[i]))
     }
 
+    pub open spec fn filter(self, pred: FnSpec(A) -> bool) -> Self
+        decreases self.len()
+    {
+        if self.len() == 0 {
+            self
+        } else {
+            let subseq = self.drop_last().filter(pred);
+            if pred(self.last()) { subseq.push(self.last()) } else { subseq }
+        }
+    }
+
+    pub proof fn filter_lemma(self, pred: FnSpec(A) -> bool)
+        ensures
+            // we don't keep anything bad
+            // TODO(andrea): recommends didn't catch this error, where i isn't known to be in
+            // self.filter(pred).len()
+            //forall |i: int| 0 <= i < self.len() ==> pred(#[trigger] self.filter(pred)[i]),
+            forall |i: int| 0 <= i < self.filter(pred).len() ==> pred(#[trigger] self.filter(pred)[i]),
+            // we keep everything we should
+            forall |i: int| 0 <= i < self.len() && pred(self[i])
+                ==> #[trigger] self.filter(pred).contains(self[i]),
+            // we also keep them in order, but I can't think of how to write
+            // that just now.
```

Comment:
The Dafny `Filter` doesn't have this property either (thanks @parno for looking it up), so I'd say it's fine as it is, until we find out it's needed. Can you remove this comment?